### PR TITLE
[WIP] bpo-1635741: Port _collections module to multiphase initialization.

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue1635741

### CODE REVIEW

File: Modules/_collectionsmodule.c [hunk 1]
```diff
@@ -2546,24 +2546,56 @@ static PyTypeObject tuplegetter_type = {
...
+    PyTypeObject *typelist[] = {
+        &deque_type,
+        &defdict_type,
+        &PyODict_Type,
+        &dequeiter_type,
+        &dequereviter_type,
+        &tuplegetter_type,
+    };
+
+    for (int i = 0; typelist[i] != NULL; i++) {
```
Comment (corona10) [inline Modules/_collectionsmodule.c:2571]: @vstinner I'd like to create a new helper function for this.

This logic can be reused.
https://github.com/python/cpython/blob/514c469719f149e1722a91a9d0c63bf89dfefb2a/Modules/itertoolsmodule.c#L4733L4747

What do you think?

Comment (vstinner) [reply Modules/_collectionsmodule.c:2571]: Do you mean combine PyType_Ready + _PyType_Name + PyModule_AddObject? Yeah, maybe it's worth it, but I would prefer to add it once this PR lands.

File: Modules/_collectionsmodule.c [hunk 2]
```diff
@@ -2546,24 +2546,56 @@ static PyTypeObject tuplegetter_type = {
...
 
+static int
+collections_exec(PyObject *m) {
+    PyTypeObject *typelist[] = {
+        &deque_type,
+        &defdict_type,
+        &PyODict_Type,
+        &dequeiter_type,
+        &dequereviter_type,
+        &tuplegetter_type,
```
Comment (vstinner) [inline Modules/_collectionsmodule.c:2568]: Either you use Py_ARRAY_COUNT() to get the list length, or you need to add a NULL terminator. Currently, you rely on an undefined behavior.

**Comment by shihai1991** [comment]
FWIW, some comments in https://github.com/python/cpython/pull/18066/files

**Pull request closed without merging**